plugins/node/opentelemetry-instrumentation-undici/src/undici.ts

```diff
@@ -0,0 +1,460 @@
+        },
+        activeCtx
+      );
+    }
+
+    // Execute the request hook if defined
+    safeExecuteInTheMiddle(
+      () => config.requestHook?.(span, request),
+      e => e && this._diag.error('caught requestHook error: ', e),
+      true
+    );
+
+    // Context propagation goes last so no hook can tamper
+    // the propagation headers
+    const requestContext = trace.setSpan(context.active(), span);
+    const addedHeaders: Record<string, string> = {};
+    propagation.inject(requestContext, addedHeaders);
+
+    request.headers += Object.entries(addedHeaders)
+      .map(([k, v]) => `${k}: ${v}\r\n`)
+      .join('');
+    this._recordFromReq.set(request, { span, attributes, startTime });
+  }
+
+  // This is the 2nd message we receive for each request. It is fired when connection with
+  // the remote is established and about to send the first byte. Here we do have info about the
+  // remote address and port so we can populate some `network.*` attributes into the span
+  private onRequestHeaders({ request, socket }: RequestHeadersMessage): void {
+    const record = this._recordFromReq.get(request as UndiciRequest);
+
+    if (!record) {
+      return;
+    }
+
+    const config = this._getConfig();
+    const { span } = record;
+    const { remoteAddress, remotePort } = socket;
+    const spanAttributes: Attributes = {
+      [SemanticAttributes.NETWORK_PEER_ADDRESS]: remoteAddress,
+      [SemanticAttributes.NETWORK_PEER_PORT]: remotePort,
+    };
+
+    // After hooks have been processed (which may modify request headers)
+    // we can collect the headers based on the configuration
+    const rawHeaders = request.headers.split('\r\n');
+    const reqHeaders = new Map(
+      rawHeaders.map(h => {
+        const sepIndex = h.indexOf(':');
+        const name = h.substring(0, sepIndex).toLowerCase();
+        const val = h.substring(sepIndex + 1).trim();
+        return [name, val];
+      })
+    );
+
+    if (config.headersToSpanAttributes?.requestHeaders) {
+      config.headersToSpanAttributes.requestHeaders
+        .map(name => name.toLowerCase())
+        .filter(name => reqHeaders.has(name))
+        .forEach(name => {
+          spanAttributes[`http.request.header.${name}`] = reqHeaders.get(name);
+        });
+    }
+
+    span.setAttributes(spanAttributes);
+  }
+
+  // This is the 3rd message we get for each request and it's fired when the server
+  // headers are received, body may not be accessible yet.
+  // From the response headers we can set the status and content length
+  private onResponseHeaders({
+    request,
+    response,
+  }: ResponseHeadersMessage): void {
+    const record = this._recordFromReq.get(request);
+
+    if (!record) {
+      return;
+    }
+
+    const { span, attributes, startTime } = record;
+    // We are currently *not* capturing response headers, even though the
+    // intake API does allow it, because none of the other `setHttpContext`
+    // uses currently do
```

Comment:
Copy pasta from elastic-apm-node?

Reply:
Actually is from one of the instrumentations referred in the [original issue](https://github.com/open-telemetry/opentelemetry-js/issues/4333)

https://github.com/gas-buddy/service/blob/main/src/telemetry/fetchInstrumentation.ts#L44

I guess this comment is out of context so it can be removed.